src/inmanta/deploy/scheduler.py

```diff
@@ -357,6 +382,43 @@ async def _run_for_agent(self, agent: str) -> None:
 
             self._work.agent_queues.task_done(agent, task)
 
+    async def is_running(self) -> bool:
+        current_environment = await Environment.get_by_id(self.environment)
+        assert current_environment
+        requested_agent = await data.Agent.get(env=self.environment, endpoint=const.AGENT_SCHEDULER_ID)
+        return not (current_environment.halted or requested_agent.paused)
+
+    async def refresh_agent_state_from_db(self, name: str) -> None:
+        """
+        Refresh from the DB (authoritative entity) the actual state of the agent.
+            - If the agent is not paused: It will make sure that the agent is running.
+            - If the agent is paused: Stop a particular agent.
+
+        :param name: The name of the agent
+        """
+        requested_agent = await data.Agent.get(env=self.environment, endpoint=name)
+
+        # This should never occur as the Scheduler is creating these 'agents'!
+        if requested_agent and name not in self._state.agent_status:
+            raise RuntimeError("Agent exists in Database but not in the Scheduler state!")
+
+        if name not in self._state.agent_status and requested_agent is None:
+            raise LookupError(f"The agent `{name}` does not exist!")
+
+        is_expected_state = requested_agent.paused == (self._state.agent_status[name] == AgentStatus.STOPPED)
+        if is_expected_state:
+            return
+
+        if requested_agent.paused:
+            # We don't need to stop it, through the executor manager, because it will taking new task from the queue,
+            # so it will time out eventually
+            self._state.agent_status[name] = AgentStatus.STOPPED
+        else:
+            old_status = self._state.agent_status[name]
+            self._state.agent_status[name] = AgentStatus.STARTED
+            if old_status != AgentStatus.STARTED and (name not in self._workers or self._workers[name].done()):
+                self._start_for_agent(name)
```

Comment:
this code path will go to the database twice, seems a bit wasteful?

Reply:
I think this may also cause an agent to not run at all. The fact that the task isn't done doesn't mean it won't fall flat on its face soon.... 


I think this code should go into a nice little object, so there is an overview.